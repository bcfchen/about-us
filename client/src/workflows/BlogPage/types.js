import PropTypes from 'prop-types';

export const propTypes = {
    posts: PropTypes.array.isRequired,
    prevButtonActive: PropTypes.bool.isRequired,
    nextButtonActive: PropTypes.bool.isRequired,
    searchText: PropTypes.string
};