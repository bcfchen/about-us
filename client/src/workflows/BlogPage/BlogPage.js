import React from "react";
import * as blogActions from "../../redux/actions/blogActions";
import toast from "toastr";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import Button from '@material-ui/core/Button';
import BlogPostList from "../../components/BlogPostList/BlogPostList";
import BlogSearch from "../../components/BlogSearch/BlogSearch";
import * as _ from 'lodash';
import { propTypes } from './types';


export class BlogPage extends React.Component {
    constructor(props) {
        super(props);
        this.state = { isLoading: true };
        this.handleSearch = this.handleSearch.bind(this);
    }

    componentDidMount() {
        this.props.blogActions.searchPostByText().then(() => {
            this.setState({ isLoading: false })
        }).catch(err => {
            toast.error(err.message);
            this.setState({ isLoading: false })
        });
    }

    handleSearch(e) { this.props.blogActions.searchPostByText(e.target.value); };

    render() {
        return (
            <div>
                <BlogSearch searchText={this.props.searchText} onSearch={this.handleSearch} />
                <BlogPostList isLoading={this.state.isLoading} postsToRender={this.props.posts} />
                <div className="btns-container">
                    <Button className="prev-btn" variant="contained" disabled={!this.props.prevButtonActive} onClick={this.props.blogActions.prevPage}>Prev</Button>
                    <Button className="next-btn" variant="contained" disabled={!this.props.nextButtonActive} onClick={this.props.blogActions.nextPage}>Next</Button>
                </div>
            </div>);
    }
}

const mapDispatchToProps = dispatch => {
    return { blogActions: bindActionCreators(blogActions, dispatch) }
};

const mapStateToProps = (state, ownProps) => {
    let blogPageState = state.blogPage;
    return {
        posts: blogPageState.posts.slice(blogPageState.startIndex, blogPageState.endIndex),
        prevButtonActive: blogPageState.prevButtonActive,
        nextButtonActive: blogPageState.nextButtonActive,
        searchText: blogPageState.searchText
    };
};

BlogPage.propTypes = propTypes;
export default connect(mapStateToProps, mapDispatchToProps)(BlogPage);