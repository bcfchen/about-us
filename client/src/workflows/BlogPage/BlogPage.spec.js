import React from "react";
import { shallow } from "enzyme";
import { BlogPage } from "./BlogPage";
import toast from "toastr";
jest.mock("toastr");

describe("BlogPage", () => {
    // setup mock properties
    let defaultProps = {
        blogActions: { searchPostByText: () => Promise.resolve() },
        posts: [],
        prevButtonActive: true,
        nextButtonActive: true,
        searchText: ""
    },
        shallowComponent = null;

    let shallowRender = (props = defaultProps) => { shallowComponent = shallow(<BlogPage {...props} />); };

    describe("on load", () => {
        it("should render correctly", () => {
            shallowRender();
            expect(shallowComponent).toMatchSnapshot();
        });

        describe("loading initial blogs", () => {
            it("should call searchPostByText to fetch data", () => {
                // given
                let mockBlogActions = {
                    searchPostByText: jest.fn(() => Promise.resolve())
                };
                let overrideProps = {
                    ...defaultProps,
                    blogActions: mockBlogActions
                };
                // when
                shallowRender(overrideProps);
                // then
                expect(mockBlogActions.searchPostByText).toHaveBeenCalled();
            });

            describe("on error fetching blog data", () => {
                it("should call toast error", () => {
                    // given
                    let overrideProps = { ...defaultProps };
                    overrideProps.blogActions.searchPostByText = jest.fn(() => Promise.reject("bad"));
                    toast.error = jest.fn();
                    shallowRender(overrideProps);

                    // when
                    expect(toast.error).toHaveBeenCalled();
                })
            });
        });
    });

    // describe("on location selected", () => {
    //     describe("if valid location is provided", () => {
    //         it("actions should be called to reload weather data", () => {
    //             // given
    //             let mockWeatherActions = {
    //                 loadFiveDayForecasts: jest.fn(),
    //                 loadHourlyForecasts: jest.fn(),
    //                 loadCurrentLocationHourlyForecasts: jest.fn()
    //             };
    //             let overrideProps = { ...defaultProps };
    //             overrideProps.weatherActions = mockWeatherActions;
    //             overrideProps.locationActions.updateCurrentLocation = jest.fn();

    //             // when
    //             shallowRender(overrideProps);
    //             shallowComponent.find("WithStyles(LocationSearch)")
    //                 .prop("onLocationSelected")({ value: "locationVal" });

    //             // then
    //             expect(mockWeatherActions.loadFiveDayForecasts).toHaveBeenCalled();
    //             expect(mockWeatherActions.loadCurrentLocationHourlyForecasts).toHaveBeenCalled();
    //         })

    //         it("should call toast error if unsuccessful", () => {
    //             // given
    //             let overrideProps = { ...defaultProps };
    //             overrideProps.locationActions.updateCurrentLocation = jest.fn(() => Promise.reject("bad"));
    //             toast.error = jest.fn();
    //             shallowRender(overrideProps);

    //             // when
    //             shallowComponent.find("WithStyles(LocationSearch)")
    //                 .prop("onLocationSelected")({ value: "locationVal" }).then(() => {
    //                     // then
    //                     expect(toast.error).toHaveBeenCalled();
    //                 });
    //         })
    //     });

    //     describe("if valid location is not provided", () => {
    //         it("actions should not be called to reload weather data", () => {
    //             // given
    //             let mockWeatherActions = {
    //                 loadFiveDayForecasts: jest.fn(),
    //                 loadHourlyForecasts: jest.fn(),
    //                 loadCurrentLocationHourlyForecasts: jest.fn()
    //             };
    //             let overrideProps = { ...defaultProps };
    //             overrideProps.weatherActions = mockWeatherActions;
    //             overrideProps.locationActions.updateCurrentLocation = jest.fn();
    //             shallowRender(overrideProps);

    //             // when
    //             shallowComponent.find("WithStyles(LocationSearch)")
    //                 .prop("onLocationSelected")(undefined);

    //             // then 
    //             expect(mockWeatherActions.loadFiveDayForecasts).toHaveBeenCalledTimes(1); // once called by constructor
    //             expect(mockWeatherActions.loadHourlyForecasts).toHaveBeenCalledTimes(0);
    //         })
    //     });
    // });

}); 