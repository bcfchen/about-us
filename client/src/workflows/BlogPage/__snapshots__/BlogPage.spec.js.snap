// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogPage on load should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BlogPage
    blogActions={
        Object {
            "searchPostByText": [Function],
          }
    }
    nextButtonActive={true}
    posts={Array []}
    prevButtonActive={true}
    searchText=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <BlogSearch
          onSearch={[Function]}
          searchText=""
/>,
        <WithStyles(BlogPostList)
          isLoading={true}
          postsToRender={Array []}
/>,
        <div
          className="btns-container"
>
          <WithStyles(Button)
                    className="prev-btn"
                    disabled={false}
                    onClick={undefined}
                    variant="contained"
          >
                    Prev
          </WithStyles(Button)>
          <WithStyles(Button)
                    className="next-btn"
                    disabled={false}
                    onClick={undefined}
                    variant="contained"
          >
                    Next
          </WithStyles(Button)>
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onSearch": [Function],
          "searchText": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "isLoading": true,
          "postsToRender": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <WithStyles(Button)
              className="prev-btn"
              disabled={false}
              onClick={undefined}
              variant="contained"
>
              Prev
</WithStyles(Button)>,
            <WithStyles(Button)
              className="next-btn"
              disabled={false}
              onClick={undefined}
              variant="contained"
>
              Next
</WithStyles(Button)>,
          ],
          "className": "btns-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Prev",
              "className": "prev-btn",
              "disabled": false,
              "onClick": undefined,
              "variant": "contained",
            },
            "ref": null,
            "rendered": "Prev",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Next",
              "className": "next-btn",
              "disabled": false,
              "onClick": undefined,
              "variant": "contained",
            },
            "ref": null,
            "rendered": "Next",
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <BlogSearch
            onSearch={[Function]}
            searchText=""
/>,
          <WithStyles(BlogPostList)
            isLoading={true}
            postsToRender={Array []}
/>,
          <div
            className="btns-container"
>
            <WithStyles(Button)
                        className="prev-btn"
                        disabled={false}
                        onClick={undefined}
                        variant="contained"
            >
                        Prev
            </WithStyles(Button)>
            <WithStyles(Button)
                        className="next-btn"
                        disabled={false}
                        onClick={undefined}
                        variant="contained"
            >
                        Next
            </WithStyles(Button)>
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onSearch": [Function],
            "searchText": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "isLoading": true,
            "postsToRender": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <WithStyles(Button)
                className="prev-btn"
                disabled={false}
                onClick={undefined}
                variant="contained"
>
                Prev
</WithStyles(Button)>,
              <WithStyles(Button)
                className="next-btn"
                disabled={false}
                onClick={undefined}
                variant="contained"
>
                Next
</WithStyles(Button)>,
            ],
            "className": "btns-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Prev",
                "className": "prev-btn",
                "disabled": false,
                "onClick": undefined,
                "variant": "contained",
              },
              "ref": null,
              "rendered": "Prev",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Next",
                "className": "next-btn",
                "disabled": false,
                "onClick": undefined,
                "variant": "contained",
              },
              "ref": null,
              "rendered": "Next",
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
